lib/logproto/logproto-proxied-text-server.c

```diff
@@ -0,0 +1,156 @@
+#include <stdio.h>
+
+#include "messages.h"
+#include "logproto-proxied-text-server.h"
+
+#define PROXY_TCP_HDR "PROXY TCP%d %s %s %d %d"
+#define PROXY_UNKNOWN_HDR "PROXY UNKNOWN"
+
+static gboolean
+_log_proto_proxied_text_server_parse_header(LogProtoProxiedTextServer *self, const guchar *msg, gsize msg_len)
+{
+    GString *input = g_string_new_len(msg, msg_len);
+    int matches = 0;
+    int compares = -1;
+
+    matches = sscanf(input->str, PROXY_TCP_HDR, &self->info->ip_version, self->info->src_ip, self->info->dst_ip,
+                     &self->info->src_port, &self->info->dst_port);
+
+    if (matches != 5)
+        compares = strcmp(input->str, PROXY_UNKNOWN_HDR);
+
+    g_string_free(input, TRUE);
+
+    if (compares == 0) {
+        msg_info("PROXY UNKOWN valid protocol header received");
+        self->info->unknown = TRUE;
+    }
+
+    return (matches == 5 || compares == 0) ? TRUE : FALSE;
+}
+
+static void
+_log_proto_proxied_text_server_add_aux_data(LogProtoProxiedTextServer *self, LogTransportAuxData *aux) {
+    gchar buf1[8];
+    gchar buf2[8];
+    gchar buf3[8];
+
+    if (self->info->unknown)
+        return;
+
+    snprintf(buf1, 8, "%i", self->info->src_port);
+    snprintf(buf2, 8, "%i", self->info->dst_port);
+    snprintf(buf3, 8, "%i", self->info->ip_version);
+
+    log_transport_aux_data_add_nv_pair(aux, "PROXIED_SRCIP", self->info->src_ip);
+    log_transport_aux_data_add_nv_pair(aux, "PROXIED_DSTIP", self->info->dst_ip);
+    log_transport_aux_data_add_nv_pair(aux, "PROXIED_SRCPORT", buf1);
+    log_transport_aux_data_add_nv_pair(aux, "PROXIED_DSTPORT", buf2);
+    log_transport_aux_data_add_nv_pair(aux, "PROXIED_IP_VERSION", buf3);
+
+    return;
+}
+
+static LogProtoStatus
+_log_proto_proxied_text_server_handshake(LogProtoServer *s)
+{
+    const guchar *msg;
+    gsize msg_len;
+    gboolean may_read = TRUE;
+    gboolean parsable;
+    LogProtoStatus status;
+
+    LogProtoProxiedTextServer *self = (LogProtoProxiedTextServer *) s;
+
+    // Fetch a line from the transport layer
+    status = self->fetch_from_buffer(&self->super.super.super, &msg, &msg_len, &may_read, NULL, NULL);
+
+    if (status != LPS_SUCCESS)
+        return status;
+
+    parsable = _log_proto_proxied_text_server_parse_header(self, msg, msg_len);
+
+    msg_debug("PROXY protocol header received", evt_tag_printf("line", "%.*s", (gint) msg_len, msg));
+
+    if (parsable) {
+        msg_info("PROXY protocol header parsed successfully");
+        return LPS_SUCCESS;
+    } else {
+        msg_error("Error parsing PROXY protocol header");
+        return LPS_ERROR;
+    }
+}
+
+static gboolean
+_log_proto_proxied_text_server_handshake_in_progress(LogProtoServer *s)
+{
+    LogProtoProxiedTextServer *self = (LogProtoProxiedTextServer *) s;
+
+    // Run handshake() only once
+    if (!self->handshake_done)
+    {
+        self->handshake_done = TRUE;
```

Comment:
I think the flag should be set in the handshake method. This method becomes a simple getter this way.